lib/galaxy/webapps/galaxy/controllers/user.py

```diff
@@ -510,21 +510,23 @@ def login( self, trans, refresh_frames=[], **kwd ):
 
     def __validate_login( self, trans, **kwd ):
         """Validates numerous cases that might happen during the login time."""
-        message = escape( kwd.get( 'message', '' ) )
         status = kwd.get( 'status', 'error' )
-        email = kwd.get( 'email', '' )
+        login = kwd.get( 'login', '' )
         password = kwd.get( 'password', '' )
-        username = kwd.get( 'username', '' )
         redirect = kwd.get( 'redirect', trans.request.referer ).strip()
         success = False
-        user = trans.sa_session.query( trans.app.model.User ).filter( trans.app.model.User.table.c.email == email ).first()
+        user = trans.sa_session.query( trans.app.model.User ).filter(or_(
+            trans.app.model.User.table.c.email == login,
+            trans.app.model.User.table.c.username == login
+        )).first()
```

Comment:
I guess you can do ``.all()`` and check len() in Python to prevent second query.